### PR TITLE
GC: optimize alloc of small object by keeping track of unused pages

### PR DESCRIPTION
**Pull request by rainers**
This simple change is what yields the larger time difference of https://github.com/dlang/druntime/pull/2520#issue-263920143, i.e. reduces execution time of the vdparser-test from approx. 10.5 s to 7.5 s.

### CODE REVIEW

File: src/gc/impl/conservative/gc.d [hunk 1]
```diff
@@ -2538,6 +2538,10 @@ struct Pool
     //  only for the first and the last page.
     uint* bPageOffsets;
 
+    // The small object pool uses the same array to keep a chain of pages with the same
+    // bin size and free pages (searchStart is first free page)
```
Comment (rainers) [inline src/gc/impl/conservative/gc.d]: This comment is a bit premature, but keeping track of pages with the same bin size will be needed for deferring incremental `recover` into later allocations.

**Comment by kinke** [comment]
Rainer on fire ;) - the GC improvements are impressive (and at the same time kinda shocking that so much potential hasn't been exploited yet).

**Comment by ZombineDev** [comment]
@rainers rebasing to master should fix the CircleCI build.

**Comment by rainers** [comment]
> reduces execution time of the vdparser-test from approx. 10.5 s to 7.5 s.

BTW: time drops to about 5 seconds when compiling with LDC!

**Comment by kinke** [comment]
Is that so surprising? 30-40% runtime reduction with LDC seems like a rough rule of thumb for general code (probably more for number-crunching code), e.g., DMD itself.

**Comment by rainers** [comment]
Not really surprising, though profiling suggested that marking is pretty much memory bandwidth limited (most of the samples were taken at the instruction reading the scanned memory). LDC compiled marking code was better similar to the overall ratio which hints that there is still more leeway for improvement.

**Comment by kinke** [comment]
[@rainers: Just as a reminder, `-disable-gc2stack` may be useful for comparisons with LDC.]

**Comment by rainers** [comment]
> [@rainers: Just as a reminder, `-disable-gc2stack` may be useful for comparisons with LDC.]

Doesn't make a notable difference. The parser doesn't create many temporary objects AFAICT.

**Pull request merged**